---
title: "DESeq2_CMCdata_Tutorial_Part1"
author: "Nickie Safarian"
date: "11/17/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is the first script of a series of R Markdown documents contaning codes for bulk brain tissue RNA-Seq analysis using DESeq2 package. 
In this script, the steps to download the common mind daataset rom synpser and preprocess it is explained.

## load the libraries

```{r, echo=FALSE, message=FALSE, include=FALSE}

# install.packages("synapser", repos = c("http://ran.synapse.org", "http://cran.fhcrc.org"))

# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("variancePartition")

```

```{r, include=FALSE, message=FALSE}

suppressPackageStartupMessages({
  library(synapser)
  synLogin('shreejoy','danfelsky')
  library(tidyverse)
  library(data.table)
  library(compositions)
  library(kableExtra)
  library(plyr)
  library(cowplot)
  library(ggpubr)
  library(variancePartition)
}) 

colorArray = colors()[grep("grey|gray", colors(), invert=TRUE)]

```

## Parse THE MPP & HBCC counts

```{r , echo=FALSE, include=FALSE}

ALL_USED_IDs = c()

# Just DLPFC counts
ALL_USED_IDs = c(ALL_USED_IDs, 'syn24172729')
counts.MPP = fread(synGet('syn24172729')$path, data.table=FALSE)
rownames(counts.MPP) = counts.MPP$Geneid
counts.MPP = counts.MPP[,-c(1:6)] 

# HBCC - DLFPC
ALL_USED_IDs = c(ALL_USED_IDs, 'syn21886235')
counts.HBCC = fread(synGet('syn21886235')$path, data.table=FALSE)
rownames(counts.HBCC) = counts.HBCC$Geneid
counts.HBCC = counts.HBCC[,-c(1:6)] 

#Merge the cohorts
geneCountsMerged = cbind(counts.MPP, counts.HBCC)

```

```{r, include=FALSE}

# save (always keep a copy of the data in your directory)
saveRDS(geneCountsMerged, "MPP_HBCC_Counts_merged.1005.Samples.Rds")

```

##  Create a download function

```{r Download_the_Data}

downloadFile = function(id, version=NULL){
  fread(synGet(id, version = version)$path, data.table = F)
}

```

## Get ancestry vector using GEMTOOLS ID 

```{r Get_Ancestry_Metadata_2, include=FALSE}

ANCESTRY_ID = 'syn17894713'
ALL_USED_IDs = c(ALL_USED_IDs, ANCESTRY_ID)
ANCESTRY.HBCC = downloadFile(ANCESTRY_ID) %>% 
  dplyr::rename(ID = 'Genotyping_Sample_ID')

ANCESTRY_ID = 'syn17346100'
ALL_USED_IDs = c(ALL_USED_IDs, ANCESTRY_ID)
ANCESTRY.MPP = downloadFile(ANCESTRY_ID) %>% 
  dplyr::rename('DNA_report..Genotyping.Sample_ID' = 'Genotyping_Sample_ID')

ANCESTRY = rbind(
  ANCESTRY.HBCC[,colnames(ANCESTRY.MPP)[-2]], 
  ANCESTRY.MPP[,colnames(ANCESTRY.MPP)[-2]]
)

```

## Get Genotype ID 

```{r, include=FALSE}

GENOTYPE_ID = "syn18358480"
ALL_USED_IDs = c(ALL_USED_IDs, GENOTYPE_ID)
GENOTYPE = downloadFile(GENOTYPE_ID) %>% 
  dplyr::select(Individual_ID, `Genotyping_Sample_ID`, `Exclude`) %>% 
  dplyr::inner_join(ANCESTRY) %>% 
  dplyr::filter(is.na(`Exclude`))

```

```{r}

#save
saveRDS(GENOTYPE, "CMC_MPP_HBCC_Genotypes.964.Samples.Rds")

```

## Get RNASeq QCmetadata 

```{r, include=FALSE}

# The CMC has two meta-data sheets
CLINICAL_ID = "syn3354385"
clinical = downloadFile(CLINICAL_ID)

ASSAY_ID = "syn24173489"
rnaseq = downloadFile(ASSAY_ID)

metadata = right_join(
  clinical,
  rnaseq,
) # this metadata has 54 columns/covariates

ALL_USED_IDs = c(ALL_USED_IDs,CLINICAL_ID, ASSAY_ID)

```

```{r}

#save
saveRDS(clinical, "CMC_Metadata_Clinical_1084.Samples.Rds")
saveRDS(rnaseq, "CMC_Metadata_RNAseq_991.Samples.Rds")
saveRDS(metadata, "CMC_Merged.clinical.&.RNAseq.Metadata_Primay_991.Samples.Rds")

```

## Choose covariates of interest in the metadata (make your metadata)

```{r, include=FALSE}

metadata = metadata %>%
  dplyr::filter(`Brain_Region` %in% c("DLPFC"))

# Classify ages older than 90 as "90+" 
metadata$`Age of Death`[metadata$`Age of Death`=="90+"] = 90

# Ensure age is numeric
metadata$`Age of Death` = as.numeric(metadata$`Age of Death`)

```


## Merge the metadata and Genotype data columns

```{r, include=FALSE}

# Merge and rename columns
METADATA = metadata %>%
  dplyr::left_join(GENOTYPE, by = c("Individual ID" = "Individual_ID")) %>%  
  dplyr::rename(
    Region = 'Brain_Region',
    PMI = 'PMI (in hours)',
    RIN = 'RIN',
    ReportExclude = 'Exclude?',
    GenotypeExclude = 'Exclude',
    LibraryBatch = 'Library_Batch',
    FlowcellBatch = 'Flowcell_Batch',
    RibozeroBatch = 'Ribozero_Batch',
    MappedReads = 'Mapped_Reads',
    IntragenicRate = 'Intragenic_Rate', 
    IntronicRate = 'Intronic_Rate', 
    IntergenicRate = 'Intergenic_Rate', 
    GenesDetected = 'Genes_Detected',
    ExpProfEfficiency = 'Expression_Profiling_Efficiency', 
    rRNARate = 'rRNA_Rate',
    TotalReads = 'Total_Reads', 
    AlignmentRate = 'Percent_Aligned', 
    TranscriptsDetected = 'Transcripts_Detected',    
    TranscriptsDetected = 'Transcripts_Detected',    
    Reported_Gender = 'Reported Gender',
    ageOfDeath = 'Age of Death',
    IndividualID = 'Individual ID') %>%
  dplyr::select(
    SampleID, IndividualID, Institution, Cohort, Reported_Gender, Sex, 
    Ethnicity, Dx, ageOfDeath, PMI, pH,  RIN, EV.1, EV.2, EV.3, EV.4, EV.5, 
    LibraryBatch, FlowcellBatch, RibozeroBatch, MappedReads, TotalReads, 
    GenesDetected, AlignmentRate, IntragenicRate, IntergenicRate, IntronicRate, 
    ExpProfEfficiency, rRNARate, TranscriptsDetected, ReportExclude, 
    GenotypeExclude)
  
 
```

## Data preprocessing
### Step 1) Data subset
```{r Preprocess the Data, include=FALSE}

# Do this to remove potential confounds
METADATA = METADATA %>%
  # Start filtering
  dplyr::filter(SampleID %in% colnames(geneCountsMerged), !is.na(SampleID)) %>%
  # Change how you consider cohort
  dplyr::mutate(Cohort = forcats::fct_recode(Cohort,
                                             `MSSM-Penn-Pitt` = 'MSSM-Penn-Pitt',
                                             `NIMH-HBCC`='NIMH-HBCC')) %>%
  # Only include Ctrl and SCZ cases
  dplyr::filter(Dx %in% c("Control")) %>% # "SCZ"
  # Only include Males and Females
  dplyr::filter(Reported_Gender %in% c("Male","Female")) %>%
  # Only include XY and XX
  dplyr::filter(Sex %in% c("XY","XX")) %>%
  # Only include white and Black folk
  dplyr::filter(Ethnicity %in% c("Caucasian","African-American"))

# Rename (as "-" is misread in R)
METADATA$Ethnicity[METADATA$Ethnicity=="African-American"] <- "African.American"


```

### Step2) Exclude NA values in these variables

```{r, include=FALSE}

ind = METADATA$SampleID[which(METADATA$ReportExclude == 1 | METADATA$GenotypeExclude)]
writeLines(paste('Following',length(ind),'samples are marked exclude'))
writeLines(paste(ind, collapse = ', '))
METADATA = METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 
ind = METADATA$SampleID [is.na(METADATA$Ethnicity) | is.na(METADATA$Institution) | is.na(METADATA$Dx)]
writeLines(paste('Following', length(ind), 'counts are missing any metadata'))
writeLines(paste(ind, collapse = ', '))
METADATA = METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 
ind = METADATA$SampleID [is.na(METADATA$PMI)]
writeLines(paste('Following', length(ind), 'counts are missing PMI'))
writeLines(paste(ind, collapse = ', '))
METADATA = METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 
ind = METADATA$SampleID [is.na(METADATA$Reported_Gender)]
writeLines(paste('Following', length(ind), 'counts are missing gender'))
writeLines(paste(ind, collapse = ', '))
METADATA = METADATA  %>% dplyr::filter(!(SampleID %in% ind)) 
ind = METADATA$SampleID [is.na(METADATA$ageOfDeath)]
writeLines(paste('Following', length(ind), 'counts are missing age of death'))
writeLines(paste(ind, collapse = ', '))
METADATA = METADATA  %>% dplyr::filter(!(SampleID %in% ind))
ind = METADATA$SampleID [is.na(METADATA$EV.1)]
writeLines(paste('Following', length(ind), 'counts are missing ancestry information'))
writeLines(paste(ind, collapse = ', '))
METADATA = METADATA  %>% dplyr::filter(!(SampleID %in% ind))
tab = table(METADATA$Individual_ID)

# drop individuals who where sequenced twice here
dropSamples = sapply( names(tab[tab > 1]), function(id){
  idx = which(METADATA$Individual_ID == id)
  df = METADATA[idx,]
  # remove individuals that have less then max reads
  i = which(df$MappedReads < max(df$MappedReads))
  rownames(METADATA)[idx[i]]
} )

#subset
METADATA = METADATA[!(rownames(METADATA) %in% dropSamples), ] # N=467                                                                 controls

```

### Step 3) Remove samples with unclear age values

```{r}

METADATA= METADATA %>% subset(! ageOfDeath <1) # removes 8 samples and 
                                               # 454 samples remain

```


### Step 4) Cell Fraction Estimates

```{r Cell Fraction Estimates}

# include estimated cell fractions in the METADATA
df_cellFractions = read.table(synGet('syn22333694')$path, row.names=1)
celFrac_ilr = ilr(df_cellFractions)
colnames(celFrac_ilr) = paste0("cellFrac_ilr_", 1:3)

METADATA = merge(METADATA, df_cellFractions, by.x = 'SampleID', by.y='row.names')
METADATA = merge(METADATA, celFrac_ilr, by.x = 'SampleID', by.y='row.names')


```

```{r}

#save
saveRDS(METADATA, "CMC_metadata_merged_with_Genotyping_CellFrac_451.Control.samples.31covar.Rds")

```


