---
title: "DESeq2_CMC_Tutorial_Part4"
author: "Nickie Safarian"
date: "11/23/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

In script 4, steps towards visualizing the selected Neuropeptides 
genes' counts and demographic features are described. The main questions is:

  - How does the expression of the select neuropeptides change with age?
   

## Load the packages

```{r}

library(tidyverse)

```


## Import the Neuropeptides list/ files

```{r}

# Vanessa's Reported NPs list
Age_NPs <- read.csv(file=file.path("/external/rprshnas01/kcni/nsafarian/CMC_Data_Analysis/Neuropeptides/Age_NPs.csv")) # 81 NPs

GO_NPs <- read.csv(file=file.path("/external/rprshnas01/kcni/nsafarian/CMC_Data_Analysis/Neuropeptides/GOSelected_NPs.csv")) # 69 NP genes

```
Note: The neuropeptide files have only gene's name, and my attempt here is to add 
ENSEMBL_gene_ID column to these files. This may help with the  process of merging Age_related_DEG file to the NPs list. 


## Add Ensembl.IDs column to the NPs' gene names 

```{r}

library( "biomaRt" )
ensembl <- useMart( "ensembl", dataset = "hsapiens_gene_ensembl" )
gene.map <- getBM(attributes = 
                    c("ensembl_gene_id", 'external_gene_name', "hgnc_symbol"),
                  filters = "hgnc_symbol", 
                  values = Age_NPs$Gene.Symbol, 
                  mart = ensembl) # 86 genes is listed, but the
                                  # Age_NPs has 81 Gene.Symbol!!! duplication issue:

# How many duplicate IDs are there?
sum(duplicated(gene.map$hgnc_symbol))  # 8 duplicated

# You may remove duplicates (I don't recommend):
# gene.map <- gene.map[!duplicated(gene.map$hgnc_symbol), ] #79 ids remain

# merge
Age_NPs.Full <- left_join(gene.map, Age_NPs, by=c('hgnc_symbol' = 'Gene.Symbol'))

# now the NPs' gene names has the Ensembl gene IDs column attached to it.

```


## Import the DESeq2 driven Age-related differentially expressed genes

```{r}

CMC_Age_DEGs <- read.csv(file=file.path("/external/rprshnas01/kcni/nsafarian/CMC_Data_Analysis/CommonMind.Data/DESeq2Object/DESeq2.results_for_Age.csv"))  

# set the rownames
CMC_Age_DEGs <- CMC_Age_DEGs %>% column_to_rownames(var="X") #27941 genes

```


## Subset DESeq2 driven Age-related differentially expressed genes for NPS

```{r}

select.genes <- CMC_Age_DEGs %>% subset(ensemblID %in% Age_NPs.Full$ensembl_gene_id) 
# 53 genes remain (Out of 81 NPs, only 53 matched with DESeq2 DEG results)
  
```


## construct the desired data matrix for visualization
In this matrix[x, y], rows will be samples and columns are grouping variables
as well as DEGs. 

### *Step 1. Import DEseq2 objects*

```{r}

dds <- readRDS(file=file.path("/external/rprshnas01/kcni/nsafarian/CMC_Data_Analysis/CommonMind.Data/DESeq2Object/dds_CMCdata_onlyControls.counts.more.than.10.Rds"))


vsd <- readRDS(file=file.path("/external/rprshnas01/kcni/nsafarian/CMC_Data_Analysis/CommonMind.Data/DESeq2Object/vst.Transformed_CMCdata_onlyControls_counts.more.than.10.Rds"))


```

### *2. Transform the count matrix and subset it for the select.genes*

```{r}

# Transform the data so samples are in rows and genes are in columns
geneCounts <- as.data.frame(t(assay(vsd)))

# Subset the transformed count matrix for only select.genes:
dt <- geneCounts[, colnames(geneCounts) %in% rownames(select.genes)] 
              # 451 Control Samples * 27941 genes 

```

### *3. Add grouping variables to the matrix*

```{r}

dt$Cohort <- dds@colData$Cohort
dt$Gender <- dds@colData$Reported_Gender
dt$PMI <- dds@colData$PMI
dt$RIN <- dds@colData$RIN
dt$Age <- dds@colData$ageOfDeath


# Rearrange, so the grouping variables are at the begining of the data
dt <- dt[, c(54: 58, 1:53)]

# Check the data
library(knitr)
kable(dt [1:6, 1:8])

# save
write.csv(dt, "CMC_vsd.data_with.variables_only.Controls_27941.DEG.csv")

```

#------------------------------------------------------------------------------

##  Q1- How does the expression of the select neuropeptides change with age?

### *Visualize the correlation between age and select.genes' expression* 

```{r}

# scatter plot for individual gene.count (across age)

ggplot(dt, aes(x=Age, y=ENSG00000175206.10)) + 
  geom_point() + 
  ggtitle("ENSG00000175206.10\n (expression across age)")+
  theme_bw()+
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size = 10),
        axis.title =element_text(size=10,face="bold"))+
  theme(plot.title = element_text(size = 8, face = "bold"))+
  scale_y_log10()

```


### *Box.Plot the expression of select.genes vs. other grouping/categorical variables (like gender, cohort, etc.)*

```{r}

# plot Function, how to make all plots in one run:

plots_fun <- function (df, x_var, y_var){
  ggplot(df, aes(x=.data[[x_var]], y=.data[[y_var]]))+
    geom_boxplot(outlier.shape=NA)+
    geom_jitter(shape=1, color="coral2", position=position_jitter(0.2))+
    ggtitle(paste0("...."), y_var)+
    labs(x= x_var, y=y_var)+
    theme(axis.text = element_text(size=10, face="bold"),
          axis.title = element_text(size=10, face="bold"))+
    theme_bw()
}


# call the function to make a list of plot 
NPs_plot_list <- colnames(dt [, -c(1:5)]) %>%  
  map (~ plots_fun (dt, colnames(dt)[1], .x))

# Note two things: 
     # 1) you need to remove the categorical variable columns before calling 
     # for the function (that's why i used "colnames(dt [, -c(1:5)])" , where 
     # column 1 to 5 are all categorical). 
     # 2) in the map() function the colnames(dt)[1] part asks to consider 
     # column 1 as the grouping variable for plotting counts. In this case it's 
     # cohorts. But for sure we can use Gender column too. 
     

# save
saveRDS(NPs_plot_list , "select_neuropeptides_plot_list.Rds")

```


Note: instead of making boxplot for expression values vs. a grouping variable, we can plot the expression levels vs. age, which of interest in this project. To make a scatter plot for those select.genes run the followings: 

```{r}
# scatter plot for NPs expression vs. Age
scatterPlot_fun <- function (df, x_var, y_var){
  ggplot(df, aes(x=.data[[x_var]], y=.data[[y_var]]))+
    geom_point()+
    ggtitle(paste0("...."), y_var)+
    labs(x= x_var, y=y_var)+
    theme(axis.text = element_text(size=10, face="bold"),
          axis.title = element_text(size=10, face="bold"))+
    theme_bw()
}


# call the function to make a list of plot 
NPs.Age_Scatterplots <- colnames(dt [, -c(1:5)]) %>%  
  map (~ scatterPlot_fun(dt, colnames(dt)[5], .x)) # column 5 is Age

# save
saveRDS(NPs.Age_Scatterplots , "select_neuropeptides_vs_Age_plot_list.Rds")

```

Note: only 26 of the GO_selected_NPs show a decline in the gene expression levels in the Common Mind dataset. 

#--------------------------------------------------------------------------------------------------------------------------

## Extra Notes (very helpful) 

vsd() function calculates a variance stabilizing transformation (VST) from the 
fitted dispersion-mean relation(s) and then transforms the count data 
(normalized by division by the size factors or normalization factors), 
yielding a matrix of values which are now approximately homoskedastic 
(having constant variance along the range of mean values). The transformation 
also normalizes with respect to library size. The rlog is less sensitive to 
size factors, which can be an issue when size factors vary widely. These 
transformations are useful when checking for outliers or as input for machine 
learning techniques such as clustering or linear discriminant analysis.


Since, some researchers prefer to show the expression levels by the log-transformed counts values,
I'm including the following line of codes so you can vidualize counts this way too: 

```{r}

# load the packages
library(edgeR)
library(limma)
library(Glimma)
library(Rcpp)

# Get log2 counts per million
CMC_CPM <- cpm(cts,log=TRUE) # remeber to load the cts data first

# convert it into a data frame
CMC_CPM <- CMC_CPM %>% as.data.frame() %>% 
  tibble::rownames_to_column(var = "ensembl_id") # 451 control samples


# remove the decimals from the end of ensembl_ids
CMC_CPM$IDs <- sapply(strsplit(CMC_CPM$ensembl_id, split="\\." ), "[", 1 )
CMC_CPM <- CMC_CPM[, c(453, 1: 452)]

# subset the CPM data for only selected neuropeptides and transform the data 
select.genes.CPM = CMC_CPM %>% 
  filter(IDs %in%  Age_NPs.Full$ensembl_gene_id) %>% 
  t() %>% as.data.frame() %>% 
  tibble::rownames_to_column(var = 'SampleID') # all 79 NP genes are found


# now make "Ensembl_ID" row as colnames
colnames(select.genes.CPM) <- select.genes.CPM[1, ] 

# # remove "ensembl_id" row (which has the decimal and sits in the first row)
select.genes.CPM <- select.genes.CPM[-(1:2), ] 
                                           
# I also checked if sample ID rows are in the same order between metadata and select.genes.CPM (it's just for extra cautions)
all(select.genes.CPM$IDs == METADATA$SampleID)


# Now Add Age column from metadata 
select.genes.CPM$Age <- METADATA$ageOfDeath # it's a 451 * 80 matrix

# reorder so Age column is at the begining of the matrix
select.genes.CPM <- select.genes.CPM[, c(1,80, 2:79)] 

# arrange data by age (ascending)
select.genes.CPM <- select.genes.CPM %>% 
  arrange(Age)

# convert all character columns to as.numeric for plotting 
select.genes.CPM[, -1] <- sapply(select.genes.CPM[, -1], as.numeric)
sapply(select.genes.CPM, class)


# plot the counts for a single gene
ggplot(select.genes.CPM, aes(x=Age, y=ENSG00000175206)) + 
  geom_point() + 
  ggtitle("NPPA expression across age)")+
  theme_bw()+
  theme(axis.text.y = element_text(size = 10),
        axis.title =element_text(size=10,face="bold"))+
  theme(plot.title = element_text(size = 8, face = "bold"))+
  scale_x_continuous(limits=c(0, 100))
  


```











